pandas/tests/io/json/test_pandas.py

```diff
@@ -1012,77 +1012,86 @@ def test_convert_dates_infer(self):
             result = read_json(dumps(data))[["id", infer_word]]
             assert_frame_equal(result, expected)
 
-    def test_date_format_frame(self):
+    @pytest.mark.parametrize(
+        "date,date_unit",
+        [
+            ("20130101 20:43:42.123", None),
+            ("20130101 20:43:42", "s"),
+            ("20130101 20:43:42.123", "ms"),
+            ("20130101 20:43:42.123456", "us"),
+            ("20130101 20:43:42.123456789", "ns"),
+        ],
+    )
+    def test_date_format_frame(self, date, date_unit):
         df = self.tsframe.copy()
 
-        def test_w_date(date, date_unit=None):
-            df["date"] = Timestamp(date)
-            df.iloc[1, df.columns.get_loc("date")] = pd.NaT
-            df.iloc[5, df.columns.get_loc("date")] = pd.NaT
-            if date_unit:
-                json = df.to_json(date_format="iso", date_unit=date_unit)
-            else:
-                json = df.to_json(date_format="iso")
-            result = read_json(json)
-            expected = df.copy()
-            expected.index = expected.index.tz_localize("UTC")
-            expected["date"] = expected["date"].dt.tz_localize("UTC")
-            assert_frame_equal(result, expected)
-
-        test_w_date("20130101 20:43:42.123")
-        test_w_date("20130101 20:43:42", date_unit="s")
-        test_w_date("20130101 20:43:42.123", date_unit="ms")
-        test_w_date("20130101 20:43:42.123456", date_unit="us")
-        test_w_date("20130101 20:43:42.123456789", date_unit="ns")
+        df["date"] = Timestamp(date)
+        df.iloc[1, df.columns.get_loc("date")] = pd.NaT
+        df.iloc[5, df.columns.get_loc("date")] = pd.NaT
+        if date_unit:
+            json = df.to_json(date_format="iso", date_unit=date_unit)
+        else:
+            json = df.to_json(date_format="iso")
+        result = read_json(json)
+        expected = df.copy()
+        # expected.index = expected.index.tz_localize("UTC")
```

Comment:
I commented this out as it was failing the test suite, but I don't think this is actually desired per #12997 . This change actually prevents index / column labels from undesirably being cast to UTC

The downside to doing this here is that how we handle timestamps differs slightly now depending on whether the timestamp appears as a JSON object key (for example, index and/or column values) versus as a value

Values get encoded via the `make_iso_8601_datetime` function that appears in tslibs/src/datetime/np_datetime_strings.c . I know @jbrockmendel was thinking about decoupling this dependency which we may be able to do when fully aligning the logic for creating a label versus encoding the data